packages/experiment-tag/src/experiment.ts

```diff
@@ -8,443 +8,675 @@ import {
 
-  setUrlChangeListener(new Set([...localFlagKeys, ...remoteFlagKeys]));
+    if (
+      !this.isRemoteBlocking ||
+      !this.config.applyAntiFlickerForRemoteBlocking
+    ) {
+      // Remove anti-flicker css if remote flags are not blocking
+      this.globalScope?.document.getElementById?.('amp-exp-css')?.remove();
+    }
 
-  // apply local variants
-  applyVariants(globalScope.webExperiment.all(), localFlagKeys);
+    if (this.remoteFlagKeys.length === 0) {
+      return;
+    }
 
-  if (!isRemoteBlocking) {
-    // Remove anti-flicker css if remote flags are not blocking
-    globalScope.document.getElementById?.('amp-exp-css')?.remove();
+    await this.fetchRemoteFlags();
+    // apply remote variants - if fetch is unsuccessful, fallback order: 1. localStorage flags, 2. initial flags
+    this.applyVariants(this.remoteFlagKeys);
   }
 
-  if (remoteFlagKeys.size === 0) {
-    return;
+  /**
+   * Get the underlying ExperimentClient instance.
+   */
+  public getExperimentClient(): ExperimentClient | undefined {
+    return this.experimentClient;
   }
 
-  try {
-    await globalScope.webExperiment.doFlags();
-  } catch (error) {
-    console.warn('Error fetching remote flags:', error);
+  /**
+   * Set the context for evaluating experiments.
+   * If user is not provided, the current user is used.
+   * If currentUrl is not provided, the current URL is used.
+   * @param user
+   * @param currentUrl
+   */
+  public setContext(
+    user: ExperimentUser | undefined = undefined,
+    currentUrl: string | undefined = undefined,
+  ) {
+    if (this.experimentClient) {
+      const existingUser = this.experimentClient?.getUser();
+      if (user) {
+        if (currentUrl) {
+          // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+          // @ts-ignore
+          user.currentUrl = currentUrl;
+        }
+        this.experimentClient.setUser(user);
+      } else {
+        this.experimentClient.setUser({
+          ...existingUser,
+          // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+          // @ts-ignore
+          currentUrl: currentUrl,
+        });
+      }
+    }
   }
-  // apply remote variants - if fetch is unsuccessful, fallback order: 1. localStorage flags, 2. initial flags
-  applyVariants(globalScope.webExperiment.all(), remoteFlagKeys);
-};
-
-const applyVariants = (
-  variants: Variants,
-  flagKeys: Set<string> | undefined = undefined,
-) => {
-  if (Object.keys(variants).length === 0) {
-    return;
+
+  /**
+   * Set the previous URL for tracking back/forward navigation. Set previous URL to prevent infinite redirection loop
+   * in single-page applications.
+   * @param url
+   */
+  public setPreviousUrl(url: string) {
+    this.previousUrl = url;
   }
-  const globalScope = getGlobalScope();
-  if (!globalScope) {
-    return;
+
+  /**
+   * Apply evaluated variants to the page.
+   * @param flagKeys
+   */
+  public applyVariants(flagKeys: string[] | undefined = undefined) {
```

Comment:
Change to object input, `ApplyVariantsOptions`, add `flagKeys` as optional field.